public/app/features/explore/LogRowContext.tsx

```diff
@@ -0,0 +1,239 @@
+  List,
+  GrafanaTheme,
+  selectThemeVariant,
+  ClickOutsideWrapper,
+  CustomScrollbar,
+  DataQueryError,
+} from '@grafana/ui';
+import { css, cx } from 'emotion';
+import { LogRowContextRows, HasMoreContextRows, LogRowContextQueryErrors } from './LogRowContextProvider';
+import { LogRowModel } from 'app/core/logs_model';
+import { Alert } from './Error';
+
+interface LogRowContextProps {
+  row: LogRowModel;
+  context: LogRowContextRows;
+  errors?: LogRowContextQueryErrors;
+  hasMoreContextRows: HasMoreContextRows;
+  onOutsideClick: () => void;
+  onLoadMoreContext: () => void;
+}
+
+const getLogRowContextStyles = (theme: GrafanaTheme) => {
+  const gradientTop = selectThemeVariant(
+    {
+      light: theme.colors.white,
+      dark: theme.colors.dark1,
+    },
+    theme.type
+  );
+  const gradientBottom = selectThemeVariant(
+    {
+      light: theme.colors.gray7,
+      dark: theme.colors.dark2,
+    },
+    theme.type
+  );
+
+  const boxShadowColor = selectThemeVariant(
+    {
+      light: theme.colors.gray5,
+      dark: theme.colors.black,
+    },
+    theme.type
+  );
+  const borderColor = selectThemeVariant(
+    {
+      light: theme.colors.gray5,
+      dark: theme.colors.dark9,
+    },
+    theme.type
+  );
+
+  return {
+    commonStyles: css`
+      position: absolute;
+      width: calc(100% + 20px);
+      left: -10px;
+      height: 250px;
+      z-index: 2;
+      overflow: hidden;
+      background: ${theme.colors.pageBg};
+      background: linear-gradient(180deg, ${gradientTop} 0%, ${gradientBottom} 104.25%);
+      box-shadow: 0px 2px 4px ${boxShadowColor}, 0px 0px 2px ${boxShadowColor};
+      border: 1px solid ${borderColor};
+      border-radius: ${theme.border.radius.md};
+    `,
+    header: css`
+      height: 30px;
+      padding: 0 10px;
+      display: flex;
+      align-items: center;
+      background: ${borderColor};
+    `,
+    logs: css`
+      height: 220px;
+      padding: 10px;
+    `,
+  };
+};
+
+interface LogRowContextGroupHeaderProps {
+  row: LogRowModel;
+  rows: Array<string | DataQueryError>;
+  onLoadMoreContext: () => void;
+  shouldScrollToBottom?: boolean;
+  canLoadMoreRows?: boolean;
+}
+interface LogRowContextGroupProps extends LogRowContextGroupHeaderProps {
+  rows: Array<string | DataQueryError>;
+  className: string;
+  error?: string;
+}
+
+const LogRowContextGroupHeader: React.FunctionComponent<LogRowContextGroupHeaderProps> = ({
+  row,
+  rows,
+  onLoadMoreContext,
+  canLoadMoreRows,
+}) => {
+  const theme = useContext(ThemeContext);
+  const { header } = getLogRowContextStyles(theme);
+
+  // Filtering out the original row from the context.
+  // Loki requires a rowTimestamp+1ns for the following logs to be queried.
+  // We don't to ns-precision calculations in Loki log row context retrieval, hence the filtering here
+  // Also see: https://github.com/grafana/loki/issues/597
+  const logRowsToRender = rows.filter(contextRow => contextRow !== row.raw);
```

Comment:
Is `contextRow` and `row.raw` of the same type?

Reply:
`contextRow` is either string or DataQueryMessage. Header is not rendered when there is an error, so this is safe. But error handling in log row context will require some <3 to make it more verbose.